### PR TITLE
[CALCITE-3363] JoinUnionTransposeRule.RIGHT_UNION should not match SEMI/ANTI join

### PR DESCRIPTION
**Pull request by jinxing64**
JoinUnionTransposeRule works by pull up union from below to top of join. Thus it should not match semi/anti join by semantics.

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/rel/rules/JoinUnionTransposeRule.java [hunk 1]
```diff
@@ -92,7 +93,8 @@ public void onMatch(RelOptRuleCall call) {
         return;
       }
     } else {
-      if (join.getJoinType().generatesNullsOnRight()) {
+      if (join.getJoinType().generatesNullsOnRight()
+          || join.getJoinType() == JoinRelType.ANTI) {
```
Comment (yanlin-Lynn) [inline core/src/main/java/org/apache/calcite/rel/rules/JoinUnionTransposeRule.java]: make sense.

File: core/src/test/java/org/apache/calcite/test/RelOptRulesTest.java [hunk 2]
```diff
@@ -1778,6 +1778,53 @@ private void basePushFilterPastAggWithGroupingSets(boolean unchanged)
...
+        .project(builder.field("DEPTNO"))
+        .union(true)
+        .build();
+    RelNode relNode = builder.push(left).push(right)
+        .antiJoin(
+            builder.call(SqlStdOperatorTable.EQUALS,
+                builder.field(2, 0, "DEPTNO"),
+                builder.field(2, 1, "DEPTNO")))
+        .project(builder.field("SAL"))
+        .build();
```
Comment (hsyuan) [inline core/src/test/java/org/apache/calcite/test/RelOptRulesTest.java]: Is it possible to write SQL directly?

Comment (chunweilei) [reply core/src/test/java/org/apache/calcite/test/RelOptRulesTest.java]: > Is it possible to write SQL directly?

+1

**Comment by jinxing64** [comment]
@hsyuan 
>Not only ANTI, but also SEMI. It should not match SEMI either.

I guess semi-join should be fine.
```
select r1.sal from emp r1 where r1.deptno in (select deptno from dept d1 where deptno < 10 union all select deptno from dept d2 where deptno > 20)

is equivalent to 

select r1.sal from emp r1 where r1.deptno  in (select deptno from dept d1 where deptno < 10)
union all
select r1.sal from emp r1 where r1.deptno  in (select deptno from dept d1 where deptno > 20)

```

File: core/src/test/java/org/apache/calcite/test/RelOptRulesTest.java (hunk 2 above)
Comment (jinxing64) [reply core/src/test/java/org/apache/calcite/test/RelOptRulesTest.java]: hmm, I tried, but failed for below two issues:
CALCITE-3366 -- currently `RelDecorrelator` doesn't support decorrelation for Union
CALCITE-3367 -- I cannot find a proper way to construct a semi-join just like SemiJoinRule, i.e. RelOptRulesTest#testSemiJoinRuleExists

I made fixes for the above two issues in https://github.com/apache/calcite/pull/1468 and https://github.com/apache/calcite/pull/1469
It's great if you can take a look when you have time

**Comment by hsyuan** [comment]
What if `deptno < 10` becomes `deptno < 100`?

**Comment by jinxing64** [comment]
hmm ~ yes, you are right 
Thanks @hsyuan

**Comment by jinxing64** [comment]
Thanks for approve @hsyuan

**Pull request closed without merging**